[Reviewer: Ellie] Allow DnsCachedResolver to load static DNS records

sebrexmetaswitch opened the pull request:
- constructors now take an optional filename
 - it will try to read this as a JSON file, storing the static records in memory
 - when doing a dns_query, these records will be examined first.

Currently, only CNAME records are supported
 - if one is found, the DNS query is performed on the redirected host rather than the original

Testing done:
 - UTs (in cpp-common-test)
 - some live testing (though more will follow)

eleanor-merry commented on src/dnscachedresolver.cpp:
```diff
@@ -207,6 +226,114 @@ DnsCachedResolver::~DnsCachedResolver()
+          }
+          else
+          {
+            TRC_DEBUG("Found unsupported record type: %s", type.c_str());
```
TRC_ERROR

eleanor-merry commented on src/dnscachedresolver.cpp:
```diff
@@ -207,6 +226,114 @@ DnsCachedResolver::~DnsCachedResolver()
+  }
+  catch (JsonFormatError err)
+  {
+    TRC_ERROR("Error parsing dns config file - records must be an array.");
```
Can you pass in the file name rather than say dns config file

sebrexmetaswitch commented on src/dnscachedresolver.cpp:
```diff
@@ -207,6 +226,114 @@ DnsCachedResolver::~DnsCachedResolver()
+
+  try
+  {
+    for (rapidjson::Value::ConstMemberIterator hosts_it = doc.MemberBegin();
```
I think this will make it invalid JSON, since the hostname is actually the name of the JSON object, and the array is the value.

sebrexmetaswitch commented on src/dnscachedresolver.cpp:
```diff
@@ -207,6 +226,114 @@ DnsCachedResolver::~DnsCachedResolver()
+  if (!fs)
+  {
+    // File doesn't exist or can't be opened, just return
+    return;
```
This is actually the expected case for almost all uses of the DNSCachedResolver, since the _dns_config_file is not specified for most instances. So I've added a DEBUG log

sebrexmetaswitch commented:
Thanks, I've implemented most of those suggestions with the exception of the two comments replied to above.

Can you take a quick look and check you're happy with the new PD log strings? Just want to check they line up with what you were envisaging.

sebrexmetaswitch commented:
As discussed, I've now:
 - changed the structure of the JSON file
 - changed DnsCachedResolver to allow it to reload the static DNS records without requiring a restart.

Could you review for me please?

eleanor-merry commented on src/dnscachedresolver.cpp:
```diff
@@ -207,6 +226,166 @@ DnsCachedResolver::~DnsCachedResolver()
+        }
+        catch (JsonFormatError err)
+        {
+          TRC_ERROR("Bad DNS record specified for hostname %s in DNS config file %s",
```
PD log please

eleanor-merry commented on src/dnscachedresolver.cpp:
```diff
@@ -122,6 +128,9 @@ void DnsCachedResolver::init(const std::vector<IP46Address>& dns_servers)
   _cache_lock = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;
 
+  // Read any static dns records from the config file
+  reload_static_records();
```
You could remove this, then have the updater run on start up. That way only those callers that want to use static records will ever call this function.
